src/widget/wtrackmenu.cpp

```diff
@@ -1627,6 +1657,181 @@ void WTrackMenu::slotClearAllMetadata() {
             &trackOperator);
 }
 
+namespace {
+
+class RemoveTrackFilesFromDiskTrackPointerOperation : public mixxx::TrackPointerOperation {
+  public:
+    QList<TrackRef> getTracksToPurge() const {
+        return mTracksToPurge;
+    }
+    QList<QString> getTracksToKeep() const {
+        return mTracksToKeep;
+    }
+
+  private:
+    mutable QList<TrackRef> mTracksToPurge;
+    mutable QList<QString> mTracksToKeep;
+
+    void doApply(
+            const TrackPointer& pTrack) const override {
+        auto trackRef = TrackRef::fromFileInfo(
+                pTrack->getFileInfo(),
+                pTrack->getId());
+        VERIFY_OR_DEBUG_ASSERT(trackRef.isValid()) {
+            return;
+        }
+        QString location = pTrack->getLocation();
+        QFile file(location);
+        if (file.exists() && !file.remove()) {
+            // Deletion failed, log warning and queue location for the
+            // Failed Deletions warning.
+            qWarning()
+                    << "Queued file"
+                    << location
+                    << "could not be deleted. Track is not purged";
+            mTracksToKeep.append(location);
+        } else {
+            // File doesn't exist or was deleted.
+            // Note: we must NOT purge every single track here since
+            // TrackDAO::afterPurgingTracks would enforce a track model update (select())
+            // So we add it to the purge queue and purge all tracks at once
+            // in slotRemoveFromDisk() afterwards.
+            mTracksToPurge.append(trackRef);
+        }
+    }
+};
+
+} // anonymous namespace
+
+void WTrackMenu::slotRemoveFromDisk() {
+    const auto trackRefs = getTrackRefs();
+    QStringList locations;
+    locations.reserve(trackRefs.size());
+    for (const auto& trackRef : trackRefs) {
+        QString location = trackRef.getLocation();
+        // TODO de-duplicate when extending this feature to Playlists
```

Comment:
Why leave a TODO instead of simply calling the function after the list has been populated?

https://doc.qt.io/qt-5/qstringlist.html#removeDuplicates

Reply:
ahem..right. Thanks for the pointer!